src/spawn.ml

```diff
@@ -42,13 +42,16 @@ module Env_win32 : Env = struct
   type t = string
 
   let of_list env =
-    let len = List.fold_left env ~init:1 ~f:(fun acc s -> acc + String.length s + 1) in
-    let buf = Buffer.create len in
-    List.iter env ~f:(fun s ->
-      Buffer.add_string buf s;
-      Buffer.add_char buf '\000');
-    Buffer.add_char buf '\000';
-    Buffer.contents buf
+    if env = [] then
+      "\000\000"
+    else
```

Comment:
Can you say a word about this change?

Reply:
The environment block is supposed to be terminated with two NUL characters - if there are any bindings, that obviously in the other branch (and did in the ANSI version too). When there are no bindings, the ANSI version was passing a pointer to the _OCaml_ string `"\000"` but that will have had an additional NUL terminator so it quietly worked. With the wide version, the issue was instantly revealed by one of the tests where `CreateProcessW` will have received two NUL _bytes_ (i.e. a NUL wchar_t) but will have then "segfaulted" reading past the end of that buffer (`CreateProcessW` actually translated that to an invalid argument, which worryingly suggests that it carried on reading random memory and instead got upset that there was no `=` character anywhere or some such😱)

I'm not sure if it's worth a comment in the code - `CreateProcess` does indicate that the block should be a series of nul-terminated strings themselves terminated by another nul, it's just hazy on the case of an empty environment (not surprisingly, really).

Reply:
Thanks for the explanation; but then isn't it wrong in the compiler?

https://github.com/ocaml/ocaml/blob/dfb7abefb4df82804703e6609d957f3dcda64d29/otherlibs/unix/unix_win32.ml#L999-L1003

Reply:
Yeah, I had that same shock thought! 🙂 Just awaiting a train, and then I’ll flex some environments…

Reply:
OK, OCaml's fine because `createprocess.c` speculatively adds a nul character. I was speculatively adding that character but not zeroing it. I've fixed it by not adding it, as I think it's nicer that the function in spawn.ml is entirely responsible for producing the UTF-8 version and then the C stub transcodes exactly that string.